feat:  added support for android_ui_automator2

Balaji028 opened the pull request:
- Implements #112 
- In `ui_automator2_accessbility_tree.py` file added functionality to make sure  the output of `driver.pagesource` is valid so that it can be parsed (some times when the page source is huge there are multiple xml declarations present in string , identified them and removed using regular expression and wrapped up all the elements under root tag . For example `<root> <hierarchy> </root>`) .
- In `parse_element` method took all the attributes from xml string and  added them to properties variable for Node for each id .
- for `find_element_by_id` , for Android since the attributes vary from IOS , in `accessbility_element.py` added new values like `androidresourceid` , `androidclass`, `androidtext` , `androidcontentdesc` , `androidbounds`.  and used them in `appium_driver.py` while constructing xpath .
- In `alumni.py` added mobile_platform string default to `Android` .
-  added tests  for `element_by_id` .

Balaji028 committed:
Support UiAutomator2

p0deje commented on alumnium/alumni.py:
```diff
@@ -23,11 +23,15 @@
 
 class Alumni:
-    def __init__(self, driver: Page | WebDriver, model: Model = None):
+    def __init__(self, driver: Page | WebDriver, model: Model = None, mobile_platform: str = "Android"):
```
I don't think we need it. We should instead attempt to check what Appium driver is passed via `driver` (iOS or Android). I believe there are some properties we can check that are available in `driver` variable...

p0deje commented on alumnium/drivers/appium_driver.py:
```diff
@@ -10,6 +10,7 @@
 
 
 from alumnium.accessibility import XCUITestAccessibilityTree
+from alumnium.accessibility.uiautomator2_accessibility_tree import UIAutomator2AccessibiltyTree
```
You can add `UIAutomator2AccessibiltyTree` to `__init__.py` so it can be imported from `alumnium.accessibility` directly.

p0deje commented on alumnium/drivers/appium_driver.py:
```diff
@@ -19,12 +20,16 @@
 
 class AppiumDriver(BaseDriver):
-    def __init__(self, driver: Remote):
+    def __init__(self, driver: Remote, mobile_platfrom: str):
```
Likewise, we should determine the platform using `driver` variable instead of an explicit flag.

p0deje commented on tests/accessibility/test_uiautomator2_accessibility_tree.py:
```diff
@@ -0,0 +1,22 @@
+
+@fixture
+def simple_tree() -> UIAutomator2AccessibiltyTree:
+    return tree("uiautomator2_accessibility_tree")
```
Can you please add a test that shows asserts the output of `to_xml()` method?

Balaji028 replied:
@p0deje  , Sir I am getting stucked while running test for `to_xml()` function . Every time when i run it it is getting failed while comparing it with xml string provided in test . I used `strip()` function to remove new lines and unwanted spaces but still it failing . Could you please help me sir how to solve this issue . 

<img width="1085" height="412" alt="image" src="https://github.com/user-attachments/assets/af1e30cf-63b2-4088-abbb-97483113d68c" />

p0deje replied:
@Balaji028 Can you run pytest with `-vv` and paste the output to gist?

Balaji028 replied:
@p0deje Sir , I have made amendments to the `to_xml()`  function , now when i run test it is failing , because of some UTF-8 text characters not getting decoded properly some unnecessary characters being interpreted when comparing it , tried to run this in colab where it is able to decode properly while comparing two xml strings but in VS code terminal i am getting differences like this . could you please suggest sir how to overcome this issue? , rest everything is working fine except this.

```
-                                        <TextView id="35" text="The Vera C. Rubin Observatory in Chile releases 
the first light images from its new 8.4-meter (28 ft) telescope." resource-id="org.wikipedia.alpha:id/horizontal_scroll_list_item_text" />
+                                        <TextView id="35" text="The Vera C. Rubin Observatory in Chile releases 
the first light images from its new 8.4-meter (28Â ft) telescope." resource-id="org.wikipedia.alpha:id/horizontal_scroll_list_item_text" />

here after (28 ft) this got added (28Â ft) got added .
```
<img width="1285" height="651" alt="image" src="https://github.com/user-attachments/assets/41a8ba65-eaf8-4de8-92ed-2b3096b6fb10" />

p0deje reviewed (commented):
I think it's a great start, let's proceed with adding `to_xml` so we can actually compare with what we have on iOS.

Balaji028 committed:
removed platformModel name from appium constructor

Balaji028 committed:
initiliazed tree with empty list instead of None

Balaji028 committed:
Fixed test and refactored to_xml function

Balaji028 committed:
fix format and merge with main

Balaji028 committed:
updated appium_driver file with including android support

p0deje commented on tests/accessibility/test_uiautomator2_accessibility_tree.py:
```diff
@@ -0,0 +1,270 @@
+                                  <FrameLayout id="97" resource-id="org.wikipedia.alpha:id/view_list_card_header">
+                                    <LinearLayout id="98">
+                                      <LinearLayout id="99">
+                                        <ImageView id="100" resource-id="org.wikipedia.alpha:id/view_card_header_image" />
```
I think it would be great to include relevant attributes such as text and content-desc if they are not empty strings, and add clickable if it's true. Without these attributes, I feel like LLM wouldn't be able to understand which elements to interact with.

Balaji028 committed:
Added attributes clickable,text,contentdescription to tree

Balaji028 commented:
@p0deje  sir , I have added the attributes `text` , `content-description` , ` clickable` while constructing tree form Node object . In test file correspondingly I have added `proper_format` helper function which ensure proper encoding and decoding of strings to UTF-8 format .I tried to run test locally it passed sir .

p0deje commented:
@Balaji028 There are some failures on CI related to UTF-8 encoding - https://github.com/alumnium-hq/alumnium/actions/runs/16695964450/job/47260134723?pr=131. Frankly, I don't think you need to touch encoding at all - I personally didn't do that in other tests.

Balaji028 committed:
removed all unnecessary encoding from tests

Balaji028 commented:
@p0deje  sir , sorry I have now  removed all the unnecessary encoding and decoding totally and  have made it clean now . The two tests passed locally . Some times while comparing `xml_output` with `expected_output` there is some issue with Unicode characters in text content attribute especially with the space character getting misinterpreted  , in order to overcome this issue , I have used `NFKC` normalization form from `unicodedata` module in tests before passing to `ui_automator2` .  Sorry sir ,previously I had unnecessarily dig deeper and added encoding and decoding sorry for that . 

<img width="875" height="305" alt="image" src="https://github.com/user-attachments/assets/60c10f70-5518-46df-acd5-674ffbbff1a3" />

Balaji028 committed:
Merge branch 'alumnium-hq:main' into uiautomator2support

Balaji028 commented:
Ok sir , working on that
